BDArmory/Armor/ModuleReactiveArmor.cs

```diff
@@ -71,12 +75,33 @@ void MakeArmorSectionArray()
 
         public void UpdateSectionScales(int sectionDestroyed = -1)
         {
-            direction = -sections[sectionsRemaining-1].up; 
+            int destroyedIndex = -1;
+            if (sectionDestroyed < 0)
+                for (int i = 0; i < sectionsCount; ++i)
```

Comment:
This repeated searching for the appropriate section to destroy can be avoided by maintaining a reverse map of indices, i.e.,
```diff
diff --git a/BDArmory/Armor/ModuleReactiveArmor.cs b/BDArmory/Armor/ModuleReactiveArmor.cs
index 36bf02ea8..03e1c60f2 100644
--- a/BDArmory/Armor/ModuleReactiveArmor.cs
+++ b/BDArmory/Armor/ModuleReactiveArmor.cs
@@ -1,2 +1,4 @@
 ﻿using UnityEngine;
+using System.Collections.Generic;
+using System.Linq;
 
@@ -19,2 +21,3 @@ public class ModuleReactiveArmor : PartModule
         int[] sectionIndexes;
+        Dictionary<int, int> sectionIndexMap = [];
 
@@ -64,2 +67,4 @@ void MakeArmorSectionArray()
             sectionIndexes.Shuffle();
+            sectionIndexMap = Enumerable.Range(0, sectionIndexes.Length).ToDictionary(i => sectionIndexes[i], i => i);
+            sectionIndexMap[-1] = -1; // Start at -1 so that the ++ later on in the while loop starts at 0.
             //sections.Shuffle(); //randomize order sections get removed
@@ -77,21 +82,38 @@ public void UpdateSectionScales(int sectionDestroyed = -1)
         {
-            int destroyedIndex = -1;
+            int destroyedIndex;
             if (sectionDestroyed < 0)
-                for (int i = 0; i < sectionsCount; ++i)
             {
-                    sectionDestroyed = sectionIndexes[i];
-                    if (sectionDestroyed > 0)
+                while (++sectionIndexMap[-1] < sectionIndexes.Length && sectionIndexes[sectionIndexMap[-1]] < 0) ; // Find the first non-destroyed section, starting from the last used index.
+                if (sectionIndexMap[-1] < sectionIndexes.Length)
                 {
-                        destroyedIndex = i;
-                        break;
-                    }
+                    destroyedIndex = sectionIndexMap[-1];
+                    sectionDestroyed = sectionIndexes[destroyedIndex];
                 }
                 else
-                for (int i = 0; i < sectionsCount; ++i)
-                {
-                    if (sectionDestroyed == sectionIndexes[i])
                 {
-                        destroyedIndex = i;
-                        break;
+                    Debug.LogError($"[BDArmory.ReactiveArmor]: Tried to destroy a random section, but no sections remain!");
+                    part.Destroy();
+                    return;
                 }
+                // for (int i = 0; i < sectionsCount; ++i)
+                // {
+                //     sectionDestroyed = sectionIndexes[i];
+                //     if (sectionDestroyed > 0)
+                //     {
+                //         destroyedIndex = i;
+                //         break;
+                //     }
+                // }
+            }
+            else
+            {
+                destroyedIndex = sectionIndexMap[sectionDestroyed];
+                // for (int i = 0; i < sectionsCount; ++i)
+                // {
+                //     if (sectionDestroyed == sectionIndexes[i])
+                //     {
+                //         destroyedIndex = i;
+                //         break;
+                //     }
+                // }
             }
```
(the `while` loop is then typically O(1)).

Reply:
I'm using a linear search as AFAIK this should be faster than a dictionary when the number of elements is low. Since I doubt there's gonna be more than ~10 or 20 ERA sections per part I've decided against using a dictionary. We could have a toggle though which switches us over to using a dictionary when the section count exceeds a threshold, that would be fairly easy to implement and allow us to be performant at any N.

Reply:
Fair enough, I wasn't aware of the performance difference for small arrays, only the O(N) vs O(1) behaviour.
I did some benchmarking and yeah, the linear search is faster than using dictionaries when the size is that small. For me the break-even point was around 50-60, which is more than enough for this (it's not really an area that is going to be called that often anyway), so I don't think we need to worry about it. Fully destroying a 30 section array only took around 1μs on my machine in any case.